lib/urireader.cpp

```diff
@@ -97,12 +149,37 @@ namespace HTTP{
       }
     }
 
+    // In case of s3 URI we prepare HTTP request with AWS authorization and rely on HTTP logic below
+    if (myURI.protocol == "s3+https" || myURI.protocol == "s3+http"){
+      // Check fallback to global credentials in env vars
+      bool haveCredentials = myURI.user.size() && myURI.pass.size();
+      if(!haveCredentials) {
+        // Use environment variables
+        char * envValue = std::getenv("S3_ACCESS_KEY_ID");
+        if(envValue == NULL) {
+          FAIL_MSG("error s3 uri without credentials. Consider setting S3_ACCESS_KEY_ID env var");
+          return false;
+        }
+        myURI.user = envValue;
+        envValue = std::getenv("S3_SECRET_ACCESS_KEY");
+        if(envValue == NULL) {
+          FAIL_MSG("error s3 uri without credentials. Consider setting S3_SECRET_ACCESS_KEY env var");
+          return false;
+        }
+        myURI.pass = envValue;
+      }
+      // Do not return, continue to HTTP case
+      if(!transformS3ToHttp(downer, myURI)) return false;
```

Comment:
Considering this function is only ~30 lines long and not used anywhere else, I think it would be better if the function body simply appeared here.

Reply:
I am really against this. Done.

Reply:
Haha, yeah, I know what you mean - I normally would agree with you, but in this specific case I think it's more readable/maintainable than a separate function would be. 😇